### PR TITLE
Generalize generalize

### PR DESCRIPTION
**Pull request by Kha**
Just the first step

### CODE REVIEW

File: library/init/meta/expr.lean [hunk 1]
```diff
@@ -111,12 +111,16 @@ meta constant expr.instantiate_vars        : expr → list expr → expr
 
 protected meta constant expr.subst : expr elab → expr elab → expr elab
 
+/-- `has_var e` returns true iff e has free variables. -/
 meta constant expr.has_var       : expr → bool
```
Comment (Kha) [inline library/init/meta/expr.lean]: @leodemoura I found this name a bit confusing tbh

File: library/init/meta/interactive.lean [hunk 2]
```diff
@@ -461,16 +461,15 @@ do e ← i_to_expr p,
...
-   let e' := tgt.replace $ λa n, if a = e then some (var n.succ) else none,
-   tgt' ← (do
-     tgt' ← to_expr ``(Π x, %%e = x → %%e'),
-     type_check tgt',
-     return tgt') <|> to_expr ``(Π x, %%e = x → %%tgt),
-   assert h tgt',
+   e ← i_to_expr p,
+   -- if generalizing fails, fall back to not replacing anything
+   tgt' ← do {
+     ⟨tgt', _⟩ ← solve_aux tgt (tactic.generalize e `_ >> target),
```
Comment (Kha) [inline library/init/meta/interactive.lean]: @digama0 In the end, emulating `kabstract` by combining `solve_aux` and `generalize` turned out to be the simpler solution.

Comment (digama0) [reply library/init/meta/interactive.lean]: What does `solve_aux` do?

Comment (Kha) [reply library/init/meta/interactive.lean]: It runs a tactic on a new goal of some type and returns both the tactic result and the proof

**Comment by Kha** [comment]
Okay, I believe I've officially run out of things to further generalize.

File: library/init/meta/interactive.lean [hunk 3]
```diff
@@ -383,10 +383,44 @@ precedence `generalizing` : 0
...
+        intron n,
+        -- now try to clear hypotheses that may have been abstracted away
+        let locals := arg.fold list.nil (λ e _ acc, if e.is_local_constant then e::acc else acc),
+        locals.mfor' (try ∘ clear),
+        pure h
+      },
+      intron (n-1),
+      e ← intro1,
+      pure (e, newvars)
+   },
```
Comment (Kha) [inline library/init/meta/interactive.lean:416]: This implementation could probably be a bit brittle with just one test case, but at least should not break any `induction` invocation that worked before.

File: library/data/bitvec.lean [hunk 4]
```diff
@@ -164,12 +164,10 @@ section conversion
...
     unfold bits_to_nat list.foldl,
-      -- the next 4 lines generalize the accumulator of foldl
-    let x := 0,
-    change _ = add_lsb x b + _,
-    generalize 0 y,
-    revert x, simp,
-    induction xs with x xs ; intro y,
+    -- generalize the accumulator of foldl
+    generalize h : 0 = x, conv in (add_lsb x b) { rw ←h }, clear h,
+    simp,
```
Comment (leodemoura) [inline library/data/bitvec.lean:169]: The new syntax looks awesome!

**Pull request closed without merging**